pipelines/controller/src/main/java/org/openmrs/analytics/DWHFilesManager.java

```diff
@@ -0,0 +1,353 @@
+/*
+ * Copyright 2020-2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openmrs.analytics;
+
+import com.google.common.base.Preconditions;
+import java.io.IOException;
+import java.nio.file.DirectoryNotEmptyException;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+import javax.annotation.PostConstruct;
+import org.apache.beam.sdk.io.FileSystems;
+import org.apache.beam.sdk.io.fs.MatchResult;
+import org.apache.beam.sdk.io.fs.MatchResult.Metadata;
+import org.apache.beam.sdk.io.fs.MatchResult.Status;
+import org.apache.beam.sdk.io.fs.ResolveOptions.StandardResolveOptions;
+import org.apache.beam.sdk.io.fs.ResourceId;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Iterables;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.scheduling.annotation.EnableScheduling;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.scheduling.support.CronExpression;
+import org.springframework.stereotype.Component;
+
+/**
+ * This class is responsible for purging the DWH snapshots when they expire. It also exposes few
+ * helper methods to access the DWH files and directories.
+ */
+@EnableScheduling
+@Component
+public class DWHFilesManager {
+
+  @Autowired private DataProperties dataProperties;
+
+  private CronExpression purgeCron;
+
+  private LocalDateTime lastPurgeRunEnd;
+
+  private volatile boolean isPurgeJobRunning;
+
+  private static final Logger logger = LoggerFactory.getLogger(DWHFilesManager.class.getName());
+
+  @PostConstruct
+  public void init() {
+    purgeCron = CronExpression.parse(dataProperties.getPurgeSchedule());
+  }
+
+  private void setLastPurgeRunStatus(PurgeRunStatus status) {
+    if (status == PurgeRunStatus.SUCCESS) {
+      lastPurgeRunEnd = LocalDateTime.now();
```

Comment:
IIUC, this means that if one run fails (for whatever reason) we will attempt to run again almost immediately (i.e., within 30 seconds), correct? I would guess that it is likely that the next run also fails and given that this is an IO bound task, this can become a performance issue. Should we increase the 30 seconds check interval to a larger value?

Reply:
Yes, you are right that if the purge job fails then immediately in the next run (which is 30 sec from now) the job will be retried. Shall I increase it to 10 mins?

Reply:
Update the periodic timer to 5 minutes.

Reply:
Sounds good, thanks.